### ISSUE TITLE
Configure sagemath organization Teams, branch protection rules

### ISSUE THREAD

**Issue by mkoeppe** [sagemath/trac-to-github#85]
As sketched in https://github.com/sagemath/trac-to-github/blob/master/docs/Migration-Trac-to-Github.md#proposed-permissions-and-protections

**Comment by mkoeppe** [comment]
@sagemath/core We should probably review membership in the "core" team https://github.com/orgs/sagemath/teams/core/members 
Any thoughts on additions / retirements of inactive "core" members?

**Comment by nthiery** [comment]
Not being very active myself these days, to say the least. I let you
judge depending on the roles of each team whether I should be there or
not.

Cheers,
					Nicolas

Le Thu, Feb 02, 2023 at 02:33:43AM -0800, Matthias Köppe a écrit :
> @sagemath/core We should probably review membership in the "core" team https://github.com/orgs/sagemath/teams/core/members 
> Any thoughts on additions / retirements of inactive "core" members?
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/sagemath/trac-to-github/issues/85#issuecomment-1413518397
> You are receiving this because you are on a team that was mentioned.
> 
> Message ID: ***@***.***>
		Nicolas
--
Nicolas M. Thiéry "Isil" ***@***.***>
http://Nicolas.Thiery.name/

**Comment by kwankyu** [comment]
Just to start a discussion: will we have this hierarchy?

1. owners - full power; keep three owners always?
3. core team - what permissions or privileges? release-manager belongs here?
4. reviewer team - what permissions or privileges?
5. ad-hoc team - what permissions or privileges?
6. member - what permissions?
7. non-member - what permissions?

Who can make a team? Or who can assign what permissions to a team?

**Comment by jasongrout** [comment]
> Not being very active myself these days, to say the least. I let you judge depending on the roles of each team whether I should be there or not.

Same for me.

**Comment by kwankyu** [comment]
This is the document about roles and permissions:

https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/repository-roles-for-an-organization

**Comment by kwankyu** [comment]
?

| Team | Role | Number | number of PRs |
| --- | --- | --- | --- |
| Owner | Admin | 3 |. |
| Core | Maintain |  | >= 100 |
| Reviewer | Write |   | >= 50 |
| Member | Triage |  | >= 10 |
| non-Member | Read | | |

**Comment by kwankyu** [comment]
This is the documentation for branch protection rules:

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches

**Comment by tobiasdiez** [comment]
Like this for the develop branch (with the correct ci workflows obviously)?
![image](https://user-images.githubusercontent.com/5037600/216635068-6647ddaa-ade3-42e5-a898-4ea090fb4928.png)

(not sure if "lock branch" is overwritten by "restrict who can push")

**Comment by mkoeppe** [comment]
"require linear history" should be off

**Comment by tobiasdiez** [comment]
why? squash and an occasional rebase merge seems good enough, or not?

**Comment by mkoeppe** [comment]
Because it's not the standing practice of this project, and it's not part of the migration to change everything.

**Comment by mkoeppe** [comment]
Only the release manager team (to be created) should be able to push to `develop` and `master`. Not part of this migration to change this.

**Comment by mkoeppe** [comment]
I've created a new milestone https://github.com/sagemath/trac-to-github/milestone/5 for proposed larger changes to the workflow like this. They will need discussion on sage-devel and with the release manager.

**Comment by tobiasdiez** [comment]
I think admins and maintainers always have the right to push (merge PRs with passing checks).

**Comment by kwankyu** [comment]
> Only the release manager team (to be created) should be able to push to `develop` and `master`. Not part of this migration to change this.

This migration should provide where to start after opening the repo for new issues/prs, though everything subject to change.

The release manager team seems to coincide with the reviewer team in my table. Right? Or is it above the reviewer team in the hierarchy?

**Comment by mkoeppe** [comment]
I think the "release manager team" would at the moment only consist of Volker. He needs to be able to merge into `develop` and `master`.

"Trusted reviewers", if possible, should be able to edit others' PRs (unless the PR authors have unchecked "allow edits by maintainers".

**Comment by dimpase** [comment]
Having only release manager pushing to `develop` would keep things as slow as with trac, as many positively reviewed but unmerged PRs are a source of conflicts. 

As is mentioned somewhere, there should be a `trunk` branch for merging positively reviewed PRs passing CI, which would also serve as a base for the PRs.

In trac model, it's @vbraun 's private branch he uses for building releases - but it should  not remain private. 

 The role of the release manager would be to take care of `trunk` not going astray, and using it to build releases in `develop` branch  (after each release `trunk` should get rebased over `develop`) - and dealing with all sorts of tricky PRs, which keep failing PRs for one or another reason.

**Comment by kwankyu** [comment]
?
```
release(master) --> trunk :--> develop-modularization (ad-hoc team)
                          :--> develop-manifold (ad-hoc team)
                          :--> develop-combinat (ad-hoc team)
                          ...
```

**Comment by tobiasdiez** [comment]
> Having only release manager pushing to `develop` would keep things as slow as with trac, as many positively reviewed but unmerged PRs are a source of conflicts.

👍 

> As is mentioned somewhere, there should be a `trunk` branch for merging positively reviewed PRs passing CI, which would also serve as a base for the PRs.

Isn't the `develop` branch functioning as this "trunk" branch? 

My proposal would be the following for now, which is pretty close to the current trac workflow but decouples the role of "releasing a new version" from the right to merge PRs and thus removes this bottleneck:
- Have a `develop` branch that serves as the base/target for all PRs, and is the "default branch"
- A group of maintainers is able to merge PRs into develop (and edit PRs, to my understanding you have to have "Maintainer" role in order to edit PRs, and this automatically gives you the rights to merge PRs if all checks are passing)
- From time to time the release manager tags the develop branch to releases a beta from it
- In order to create a stable release, the release manager pushes the develop branch into master/main

Of course, there needs to be a bit of communication between the release manager and the maintainers so that "experimental" PRs are not merged directly before a stable release.

**Comment by vbraun** [comment]
We should first work on our CI game before changing the merge/release process. PRs should only be merged if they pass on the supported platforms, and right now very few people have a way of checking that. We also might want to nararow down what we support (e.g. is anyone still using 32-bit?). But thats a whole different question for when the github transition is done.

**Comment by dimpase** [comment]
I agree that before the CI is improved, we should stick to the workflow we have now - i.e. positively reviewed PRs are only merged by the release manager. 

Let's at least keep it so until release 9.8 is out the door.

**Comment by mkoeppe** [comment]
For PRs that only modify `src/`, I'd say that the current "Build & Test" workflow is sufficient. 
I think it's very rare that such a PR introduces a portability problem, including doctest problems on 32 bits.

**Comment by mkoeppe** [comment]
For now I have simply locked `master` and `develop`. Admins can still push.

**Comment by tobiasdiez** [comment]
Can someone of the admins please setup a Triage team that then can label issues/PRs.

**Comment by dimpase** [comment]
> Can someone of the admins please setup a Triage team that then can label issues/PRs.

I've promoted you to Owner. Please go ahead and create such a team

**Comment by tobiasdiez** [comment]
I've now created a Triage team (with the Triage role): https://github.com/orgs/sagemath/teams/triage/members But who can join that team and how do we organize this? @kwankyu proposed a limit of 10 PRs, which I agree is a good quote. The triage role doesn't give you too much rights, but you do can create a bit of damage by mass-changing labels or closing issues. Should we open an issue in the sage repo (pinned?) where people can reply to request membership?

I've also created a branch protection rule for "*", i.e. all other branches then master and develop, that prevent someone to create new branches. (I've tested in https://github.com/sagemath/sage/pull/34975 to create a new branch in the sage repo)

**Comment by dimpase** [comment]
isn't opening a PR creating a branch? in fact, it does, so I'm a bit worried you blocked opening PRs along the way.

**Comment by dimpase** [comment]
Atm it's quite silly that one cannot add labels to issues and prs without being on the triage team.

**Comment by kwankyu** [comment]
I wonder how fine control is possible for basic permissions of members under our free GitHub license.

**Comment by tobiasdiez** [comment]
> isn't opening a PR creating a branch? in fact, it does, so I'm a bit worried you blocked opening PRs along the way.

No, you normally push to your fork and then create a PR from the branch in your fork. But the branch protection rule only applies to branches created in the main sage repo and not in the fork. (As I've just discovered one can still create branches in the main repo as admin, but then cannot push to them).


> I wonder how fine control is possible for basic permissions of members under our free GitHub license.

It's not possible to use any customized permission rule. This is a feature restricted to Enterprise Cloud users: https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/repository-roles-for-an-organization

**Comment by dimpase** [comment]
one can probably create a GH Action to set up labels based on values in a comment

**Comment by kwankyu** [comment]
> of course, PR is created on a fork, but it leads to creation of a specially named branch on the base repo.

As I understand it, no branch is created on the base repo from a PR.

**Comment by dimpase** [comment]
> > of course, PR is created on a fork, but it leads to creation of a specially named branch on the base repo.
> 
> As I understand it, no branch is created on the base repo from a PR.

it is created, its name is `pull/ID/head`. E.g. the branch of https://github.com/sagemath/sage/pull/34970
can be obtained as follows:
```
$ git fetch origin pull/34970/head
remote: Enumerating objects: 42, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 42 (delta 30), reused 38 (delta 30), pack-reused 0
Unpacking objects: 100% (42/42), 20.78 KiB | 259.00 KiB/s, done.
From github.com:sagemath/sage
 * branch                  refs/pull/34970/head -> FETCH_HEAD

**Comment by dimpase** [comment]
On trac, only a select group (TracAdmins + few others, about 20 ppl?) was able to close tickets, AFAIK.

One can use something like https://github.com/github/issue-labeler for labelling...

**Comment by roed314** [comment]
As a short term fix, I think everyone who is a recognizable contributor (e.g. the people who were asking about labels in Matthias' talk) should be added to the Triage team.  We can then talk about whether to expand these permissions further.

**Comment by roed314** [comment]
@dimpase the issue-labeler app you pointed to means that we wouldn't actually get to use the graphical interface for adding labels, which I think is raising the barrier for contributors.

**Comment by saraedum** [comment]
@dimpase I think it's easier to just have a bot reopen issues if somebody closes them who's not supposed to do that.

**Comment by saraedum** [comment]
>  Any thoughts on others that should be added to "core" and on retirements of inactive "core" members?

Should we add everybody who was a trac admin as a start? (I took the liberty to add David Roe so I am not the only one at Sage Days who can do things.)

**Comment by dimpase** [comment]
the power of triage users will go up if we start using Discussions: https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions
they can convert issues to discussions, and delete discussions.

otherwise, I'd say add to triage team everyone with at least some contribution to sagemath.

**Comment by saraedum** [comment]
> the power of triage users will go up if we start using Discussions: https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions
they can convert issues to discussions, and delete discussions.

That also does not sound like a big problem to me. But maybe that's something to discuss again when we enable discussions.

**Comment by saraedum** [comment]
It turns out that "Triage" is not a possible base permissions for all organization members. So we'd have to explicitly add people to the Triage team.

Let me see if I can cook up a script to synchronize all members to the Triage team (and not run the script yet.)

Then again, what's the process to decide these things here? We just wait for a few people to agree and we can still revert if somebody thinks it's a bad idea?

**Comment by dimpase** [comment]
I'm just a bit worried about old inactive accounts in Triage team... I'd not just add everyone, yet.

**Comment by saraedum** [comment]
I see. My idea would be to just replicate things as they were on trac since that should not be controversial in principle. (And then tweak things later.)
Also, the interface to add people to teams is very annoying for adding lots of people. (So writing a script makes this much less annoying…)

**Comment by mezzarobba** [comment]
Besides the issues with labels, am I correct that, at the moment, **only Owners can create PR reviews**? At least I can't seem to find how to do it. I believe anyone with a github account used to be able to review trac tickets. I don't know if we still want to allow reviews from people outside the sagemath org, but I think we do want all org members to be able to do it, it it should be done as soon as possible to allow development to go on normally.

**Comment by mezzarobba** [comment]
For the benefit of people following the discussion by email (do you get notified about edits to existing comments?): I was mistaken, we *can* submit reviews already (though we are not allowed to *request* reviews).

**Comment by saraedum** [comment]
To get things going and be able to work at the Sage Days we added a manually curated (and unfortunately somewhat biased) list of 86 developers to the Triage team.

**Comment by kwankyu** [comment]
I think that the number of owners should be  bare minimum, 2 or 3 as an owner is destructively powerful. I propose: @williamstein @mkoeppe @dimpase 

and all other current "owners" should be down-graded to "core members" with Admin (or Maintain) role.

**Comment by saraedum** [comment]
Personally, I would not be afraid of our owners. It's hard to do something destructive accidentally. I'd rather have a larger group of people with privileges, similar to the current trac admins. If we are afraid that accounts might get taken over by bad actors, we could ask the owners to enable 2FA to be on the safe side.

**Comment by mezzarobba** [comment]
> the power of triage users will go up if we start using Discussions: https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions they can convert issues to discussions, and delete discussions.
> 
> otherwise, I'd say add to triage team everyone with at least some contribution to sagemath.

Apparently, when you convert an issue to a discussion, the issue is locked, not deleted. Github doesn't let you reopen the issue while the discussion exists, but it does after the discussion is deleted.

**Comment by kwankyu** [comment]
> Personally, I would not be afraid of our owners. It's hard to do something destructive accidentally. I'd rather have a larger group of people with privileges, similar to the current trac admins. If we are afraid that accounts might get taken over by bad actors, we could ask the owners to enable 2FA to be on the safe side.

It is not only about accidents. I think that important changes about sage organization/repos should be made after open discussions. Restricting the powers to make the changes to a small number of people will reduce spontaneous changes. 

Not that I observed spontaneous changes, and I am not familiar with GItHub permissions system. So my concern may be just imaginary. Anyway, it is not urgent.
